drivers/ds1307/ds1307.c

```diff
@@ -0,0 +1,208 @@
+                        size_t size);
+
+static uint8_t _convert_12_to_24(uint8_t hour)
+{
+    if (hour & DS1307_REG_HOUR_12H) {
+        uint8_t tmp = bcd_to_byte(hour & DS1307_REG_HOUR_12H_MASK);
+        if (hour & DS1307_REG_HOUR_PM) {
+            if (tmp < 12) {
+                tmp += 12;
+            }
+        }
+        else {
+            if (tmp == 12) {
+                tmp = 0;
+            }
+        }
+        hour = (bcd_from_byte(tmp) & DS1307_REG_HOUR_24H_MASK);
+    }
+    return hour;
+}
+
+int ds1307_init(ds1307_t *dev, const ds1307_params_t *params)
+{
+    int res;
+    uint8_t hour;
+
+    dev->i2c = params->i2c;
+    dev->addr = params->addr;
+
+    i2c_acquire(dev->i2c);
+    i2c_init_master(dev->i2c, params->clk);
+    /* normalize hour format */
+    res = i2c_read_reg(dev->i2c, dev->addr, DS1307_REG_HOUR, &hour);
+    if (res < 0) {
+        i2c_release(dev->i2c);
+        DEBUG("ds1307: Error reading HOUR register on init: %i\n", res);
+        return -1;
+    }
+    res = i2c_write_reg(dev->i2c,dev->addr, DS1307_REG_HOUR,
+                        _convert_12_to_24(hour));
+    i2c_release(dev->i2c);
+
+    if (res < 0) {
+        DEBUG("ds1307: Error writing HOUR register on init: %i\n", res);
+        return -1;
+    }
+    dev->nvram.read = _nvram_read;
+    dev->nvram.write = _nvram_write;
+    dev->nvram.size = (DS1307_REG_RAM_LAST - DS1307_REG_RAM_FIRST) + 1;
+    dev->nvram.extra = dev;
+    return 0;
+}
+
+int ds1307_set_time(const ds1307_t *dev, const struct tm *time)
+{
+    uint8_t regs[DS1307_REG_YEAR - DS1307_REG_SEC + 1];
+    int res;
+
+    regs[DS1307_REG_SEC] = (bcd_from_byte(time->tm_sec) & DS1307_REG_SEC_MASK);
+    regs[DS1307_REG_MIN] = (bcd_from_byte(time->tm_min) & DS1307_REG_MIN_MASK);
+    regs[DS1307_REG_HOUR] = (bcd_from_byte(time->tm_hour) &
+                                DS1307_REG_HOUR_24H_MASK);
+    regs[DS1307_REG_DOW] = (bcd_from_byte(time->tm_wday + DS1307_DOW_OFFSET) &
+                                DS1307_REG_DOW_MASK);
+    regs[DS1307_REG_DOM] = (bcd_from_byte(time->tm_mday) & DS1307_REG_DOM_MASK);
+    regs[DS1307_REG_MON] = (bcd_from_byte(time->tm_mon + DS1307_MON_OFFSET) &
+                                DS1307_REG_MON_MASK);
+    regs[DS1307_REG_YEAR] = bcd_from_byte(time->tm_year + DS1307_YEAR_OFFSET);
+    i2c_acquire(dev->i2c);
+    res = i2c_write_regs(dev->i2c, dev->addr, DS1307_REG_SEC, regs,
+                         sizeof(regs));
+    DEBUG("ds1307: wrote bytes %02x %02x %02x %02x %02x %02x %02x to device (result: %i)\n",
+          regs[DS1307_REG_SEC], regs[DS1307_REG_MIN], regs[DS1307_REG_HOUR],
+          regs[DS1307_REG_DOW], regs[DS1307_REG_DOM], regs[DS1307_REG_MON],
+          regs[DS1307_REG_YEAR], res);
+    i2c_release(dev->i2c);
+    return (res < 0) ? -1 : 0;
+}
+
+int ds1307_get_time(const ds1307_t *dev, struct tm *time)
+{
+    uint8_t regs[DS1307_REG_YEAR - DS1307_REG_SEC + 1];
+    int res;
```

Comment:
But then I need a second release call... Extra variable just might be compiled to a register use, extra function call is always more ROM.